modules/function/src/main/java/com/opengamma/strata/function/marketdata/MarketDataRatesProvider.java

```diff
@@ -122,19 +128,34 @@ public FxForwardRates fxForwardRates(CurrencyPair currencyPair) {
   //-------------------------------------------------------------------------
   @Override
   public IborIndexRates iborIndexRates(IborIndex index) {
-    return marketData.getValue(IborIndexRatesKey.of(index));
+    Curve curve = marketData.getValue(IborIndexCurveKey.of(index));
+    if (curve == null) {
+      throw new IllegalArgumentException("Unable to find forward curve: " + index);
+    }
+    return IborIndexRates.of(index, getValuationDate(), curve, timeSeries(index));
   }
 
   //-------------------------------------------------------------------------
   @Override
   public OvernightIndexRates overnightIndexRates(OvernightIndex index) {
-    return marketData.getValue(OvernightIndexRatesKey.of(index));
+    Curve curve = marketData.getValue(OvernightIndexCurveKey.of(index));
+    if (curve == null) {
+      throw new IllegalArgumentException("Unable to find forward curve: " + index);
+    }
+    return OvernightIndexRates.of(index, getValuationDate(), curve, timeSeries(index));
   }
 
   //-------------------------------------------------------------------------
   @Override
   public PriceIndexValues priceIndexValues(PriceIndex index) {
-    return marketData.getValue(PriceIndexValuesKey.of(index));
+    Curve curve = marketData.getValue(PriceIndexCurveKey.of(index));
+    if (curve == null) {
+      throw new IllegalArgumentException("Unable to find forward curve: " + index);
+    }
+    if (!(curve instanceof InterpolatedNodalCurve)) {
```

Comment:
Doing so has knock on effects as the key implements `CurveKey`. There may be other implementations of the interface that do accept `Curve`, so lets leave this for now.